libc/src/sys/auxv/linux/getauxval.cpp

```diff
@@ -0,0 +1,198 @@
+      : ptr(mmap(nullptr, AUXV_MMAP_SIZE, PROT_READ | PROT_WRITE,
+                 MAP_PRIVATE | MAP_ANONYMOUS, -1, 0)) {}
+  ~AuxvMMapGuard() {
+    if (ptr != MAP_FAILED) {
+      munmap(ptr, AUXV_MMAP_SIZE);
+    }
+  }
+  void submit_to_global() {
+    // atexit may fail, we do not set it to global in that case.
+    int ret = __cxa_atexit(
+        [](void *) {
+          munmap(auxv, AUXV_MMAP_SIZE);
+          auxv = nullptr;
+        },
+        nullptr, nullptr);
+
+    if (ret != 0)
+      return;
+
+    auxv = reinterpret_cast<AuxEntry *>(ptr);
+    ptr = MAP_FAILED;
+  }
+  bool allocated() { return ptr != MAP_FAILED; }
+};
+
+struct AuxvFdGuard {
+  int fd;
+  AuxvFdGuard() : fd(open("/proc/self/auxv", O_RDONLY | O_CLOEXEC)) {}
+  ~AuxvFdGuard() {
+    if (fd != -1) {
+      close(fd);
+    }
+  }
+  bool valid() { return fd != -1; }
+};
+
+static void initialize_auxv_once(void) {
+  // if we cannot get atexit, we cannot register the cleanup function.
+  if (&__cxa_atexit == nullptr)
+    return;
+
+  AuxvMMapGuard mmap_guard;
+  if (!mmap_guard.allocated())
+    return;
+  auto *ptr = reinterpret_cast<AuxEntry *>(mmap_guard.ptr);
+
+  // We get one less than the max size to make sure the search always
+  // terminates. MMAP private pages are zeroed out already.
+  size_t available_size = AuxvMMapGuard::AUXV_MMAP_SIZE - sizeof(AuxEntryType);
+#if defined(PR_GET_AUXV)
+  int ret = prctl(PR_GET_AUXV, reinterpret_cast<unsigned long>(ptr),
+                  available_size, 0, 0);
+  if (ret >= 0) {
+    mmap_guard.submit_to_global();
+    return;
+  }
+#endif
+  AuxvFdGuard fd_guard;
+  if (!fd_guard.valid())
+    return;
+  auto *buf = reinterpret_cast<char *>(ptr);
+  libc_errno = 0;
+  bool error_detected = false;
+  while (available_size != 0) {
+    ssize_t bytes_read = read(fd_guard.fd, buf, available_size);
+    if (bytes_read <= 0) {
+      if (libc_errno == EINTR)
+        continue;
+      error_detected = bytes_read < 0;
+      break;
+    }
+    available_size -= bytes_read;
+  }
+  if (!error_detected) {
+    mmap_guard.submit_to_global();
+  }
+}
+
+static AuxEntry read_entry(int fd) {
+  AuxEntry buf;
+  ssize_t size = sizeof(AuxEntry);
+  while (size > 0) {
+    ssize_t ret = read(fd, &buf, size);
+    if (ret < 0) {
+      if (libc_errno == EINTR)
+        continue;
+      buf.id = AT_NULL;
+      buf.value = AT_NULL;
+      break;
+    }
+    size -= ret;
+  }
+  return buf;
+}
+
+LLVM_LIBC_FUNCTION(unsigned long, getauxval, (unsigned long id)) {
+  // Fast path when libc is loaded by its own initialization code. In this case,
+  // app.auxv_ptr is already set to the auxv passed on the initial stack of the
+  // process.
+  AuxvErrnoGuard errno_guard;
+
+  auto search_auxv = [&errno_guard](AuxEntry *auxv,
+                                    unsigned long id) -> AuxEntryType {
+    for (auto *ptr = auxv; ptr->id != AT_NULL; ptr++) {
+      if (ptr->id == id) {
+        return ptr->value;
+      }
+    }
```

Comment:
https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements